src/charm.py

```diff
@@ -158,48 +71,41 @@ def _on_publish_relation_changed(self, event):
-            if self.model.config[key] != self._stored.config[key]:
-                value = self.model.config[key]
-                logger.info("Setting {} to: {}".format(key, value))
-                self._stored.config[key] = value
-
-        # this should go into install hook
-        uid = pwd.getpwnam("root").pw_uid
-        gid = grp.getgrnam("root").gr_gid
-
-        path = "/etc/nginx/ssl"
-        if not os.path.exists(path):
-            os.makedirs(path, 0o755)
-            os.chown(path, uid, gid)
-            os.chmod(path, 0o755)
-
-        config = self._stored.config
-        self._stored.config["ssl_enabled"] = (
-            "ssl_cert" in config.keys()  # noqa: W503
-            and "ssl_key" in config.keys()  # noqa: W503
-            and "ssl_cert" in self._stored.config.keys()  # noqa: W503
-            and "ssl_key" in self._stored.config.keys()  # noqa: W503
-        )
-
-        # write ssl certificate if present
-        if "ssl_cert" in config.keys() and self._stored.config["ssl_cert"]:
-            cert_path = os.path.join("/etc/nginx/ssl", "server.crt")
-            cert = b64decode(self._stored.config["ssl_cert"])
-            write_file(cert_path, cert, "root", "root", 0o644)
-
-        if "ssl_key" in config.keys() and self._stored.config["ssl_key"]:
-            key_path = os.path.join("/etc/nginx/ssl", "server.key")
-            key = b64decode(self._stored.config["ssl_key"])
-            write_file(key_path, key, "root", "root", 0o640)
-
-        if "ssl_ca" in config.keys() and self._stored.config["ssl_ca"]:
-            ca_cert = self._stored.config["ssl_ca"]
-            install_ca_cert(b64decode(ca_cert), "nginx-server.crt")
+    def _on_config_changed(self, _):  # noqa: C901
+        config = self.model.config
+        for key in config:
+            self._stored.config[key] = config[key]
+
+        ssl_cert_path = "/etc/nginx/ssl/server.crt"
+        if config.get("ssl_cert"):
+            atomic_write_root_file(ssl_cert_path, b64decode(config["ssl_cert"]), 0o644)
+        else:
+            try:
+                os.remove(ssl_cert_path)
+            except FileNotFoundError:
+                pass
+
+        ssl_key_path = "/etc/nginx/ssl/server.key"
+        if config.get("ssl_key"):
+            atomic_write_root_file(ssl_key_path, b64decode(config["ssl_key"]), 0o640)
+        else:
+            try:
+                os.remove(ssl_key_path)
+            except FileNotFoundError:
+                pass
+
+        if config.get("ssl_ca"):
+            ca_cert = config["ssl_ca"]
+            atomic_write_root_file(CA_CERT_PATH, ca_cert, 0o444)
+        else:
+            try:
+                os.remove(CA_CERT_PATH)
+            except FileNotFoundError:
+                pass
+        try:
+            subprocess.check_call(["update-ca-certificates", "--fresh"])
+        except subprocess.CalledProcessError:
+            self.model.unit.status = BlockedStatus("Failed to update CA certificates")
```

Comment:
Removed install_ca_certificate() since it is only used here, add logic to remove ca cert file if config doesn't exist. Move the call of "update-ca-certificates" out of the if scope. Block the charm if it fails.
Not sure if this is the best practice.